File: docs/test-cases/graphql-security.js
// GraphQL security — intentionally vulnerable. DO NOT deploy.
const { ApolloServer } = require("@apollo/server");
const { startStandaloneServer } = require("@apollo/server/standalone");

const typeDefs = `#graphql
  type User {
    id: ID!
    name: String!
    posts: [Post!]!
    friends: [User!]!
  }
  type Post {
    id: ID!
    body: String!
    author: User!
    comments: [Comment!]!
  }
  type Comment {
    id: ID!
    text: String!
    author: User!
  }
  type Query {
    user(id: ID!): User
    users: [User!]!
  }
  type Mutation {
    login(email: String!, password: String!): String
  }
`;

// BUG: no depth limit — attacker sends { user { friends { friends { friends { ... } } } } }
// BUG: no cost analysis — alias batching bypasses rate limits
// BUG: introspection enabled in production — full schema exposed
const server = new ApolloServer({
  typeDefs,
  resolvers: {},
  introspection: true, // should be false in production
  // Missing: depthLimit, costAnalysis, alias limits
});

startStandaloneServer(server, { listen: { port: 4000 } });

File: docs/test-cases/graphql-security.py
"""GraphQL security — intentionally vulnerable. DO NOT deploy."""
import strawberry
from strawberry.fastapi import GraphQLRouter
from fastapi import FastAPI

@strawberry.type
class User:
    id: int
    name: str
    @strawberry.field
    def friends(self) -> list["User"]:
        return []  # in real code, this would query the DB
    @strawberry.field
    def posts(self) -> list["Post"]:
        return []

@strawberry.type
class Post:
    id: int
    body: str
    @strawberry.field
    def author(self) -> User:
        return User(id=1, name="test")
    @strawberry.field
    def comments(self) -> list["Comment"]:
        return []

@strawberry.type
class Comment:
    id: int
    text: str

@strawberry.type
class Query:
    @strawberry.field
    def user(self, id: int) -> User:
        return User(id=id, name="test")

# BUG: no depth limit, no cost analysis, introspection enabled
schema = strawberry.Schema(query=Query)

app = FastAPI()
app.include_router(GraphQLRouter(schema), prefix="/graphql")
